api/v1alpha3/vspherefailuredomain_types.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// VSphereFailureDomainSpec defines the desired state of VSphereFailureDomain
+type VSphereFailureDomainSpec struct {
+	// name of the Region
+	// if is defined with tags, then should be the tag name, for eg: k8s-region-us-west
+	// if not defined with tags for the case with hostgroup, then it’s a user defined region name
+	Region string `json:"region"`
+
+	// name of the Zone,
+	// if is defined with tags, then should be the tag name, for eg:  k8s-zone-us-west-az1
+	// if mapped to hostgroup, then should be the name of the hostgroup
+	Zone string `json:"zone"`
+
+	// Labels is the Tag category used to configure Region/Zone tags
+	// for FailureDomainDatacenter and FailureDomainComputeCluster
+	// +optional
+	Labels LabelsConfig `json:"labels,omitempty"`
+
+	// The underlying infrastructure for this failure domain
+	// Datacenter as the failure domain
+	Datacenter string `json:"datacenter"`
+
+	// ComputeCluster as the failure domain
+	// +optional
+	ComputeCluster *string `json:"computeCluster,omitempty"`
+
+	// HostGroup as the failure domain
+	// +optional
+	HostGroup *FailureDomainHostGroup `json:"hostGroup,omitempty"`
+
+	// the placement constraints which is used within this failure domain
+	PlacementConstaint PlacementConstraint `json:"placementConstraint"`
```

Comment:
typo: PlacementConstaint should be PlacementConstraint

Reply:
The flattening also makes the FailureDomain :  PlacementConstraint be a 1:1. So it won't be possible to make VMs within the same FD but with different constraints like place them in different folders?

Reply:
this is the only point I'm still thorn about. we need to balance UX vs the right API. I think we might want to come back to the `VSphereDeploymentZone` CRD